### ISSUE TITLE
tracker: 3D Tiles

### ISSUE THREAD

**Issue by ibgreen** [visgl/loaders.gl#1245]
Continuing from #164

## 3DTileLayer

[Sandbox to test tilesets in Cesium](https://codesandbox.io/embed/static-kmd50)

Audit
- [ ] `Tile3DLayer.loadOptions` seems strange, see #530. `tilesetOptions` would be more appropriate.

3d tiles deck.gl Example
- [ ] Clean up 3d-tiles example: Stats Widget Handling **P1** @dryabinin-actionengine  P1
- [x] Add description field support to `ion.js` and example  P1

3d-tiles debugging
- [ ] Normalize tile3d loaders protocol (@xintongxia) **P0**
- [ ] Handle region bounding boxes correctly (@xintongxia) **P0**
- [ ] Enable debug BoundingBox view (@xintongxia) **P0**
- [ ] Port region related classes to math.gl (@ibgreen / @xintongxia) **P0**
- [ ] Show view frustum in a minimap (@xintongxia/@ibgreen) **P0**

3d-tiles module documentation
- [ ] Update documentation of `Tileset3D` class **P0** (@ibgreen)
- [ ] Update documentation of generic "3d-tiles loader category" **P0** (@xintongxia / @ibgreen)

3d-tiles module tests
- [ ] Add render tests for 3d tiles (@jianhuang01) **P0**
- [ ] Add traversal tests (@xintongxia) **P0**
- [ ] pointcloud: Enable Draco related unit tests #244 (@xintongxia)  P1
- [ ] pointcloud: Enable all unit tests (@xintongxia) P1
- [ ] i3dm/b3dm: Enable all unit tests (@xintongxia) P1

glTF-based Tile parsing
- [x] Fix async loading issue with `ScenegraphLayer` **P0** 
- [ ] Fix ADDITIVE traversal (keep parent until children load) #384  P0

deck.gl
- [ ] Feature: Support fixed frame Coordinate system in deck.gl **P2**
- [x] Feature: Support unlit material  in deck.gl 

## Full compliance with 3D tiles standard
- [ ] Styles (partially ported in WIP dir)
- [ ] Composite tiles (just render the subtiles)
- [ ] Hierarchical batch table (partially ported in live code)
- [ ] Support oct-encoded normals on GPU (need shader module)
- [ ] Vector tiles (Optional tile format)
- [ ] Geometry tiles (Optional tile format)

### glTF THREE.js Example
- [ ] Refactor to build THREE.js geometries on `onLoad`
- [ ] Swap out custom traversal with loaders.gl traversal

**Comment by ibgreen** [comment]
Implemented the lazy implicit-subtree tranche in draft PR #3540 (stacked on #3528).

This tranche:
- removes subtree network I/O from root tileset parsing and `onTilesetLoad`;
- requests one subtree boundary only after visibility, viewer-request-volume, and SSE eligibility;
- routes subtree metadata through normal progressive/foveated scheduling, resolver/query inheritance, in-flight deduplication, and a bounded parsed-subtree LRU;
- preserves REPLACE coverage while availability is pending;
- fixes breadth-first tile availability, direct child-subtree Morton indexing, deep coordinate arithmetic, and `availableLevels` last-level semantics;
- adds child lifecycle and `Tiles3DSource.getImplicitTilingStats()` diagnostics;
- adds focused shared tests, a 1,365-tile benchmark, full TSDoc, an Implicit docs tab/guide, cross-links, and five What’s New bullets.

Local validation is green: build, lint, 397 Node test files / 1,728 tests, 401 headless files / 1,731 tests, and the production website build.

#3540 fixes #3291 and #3440 when merged and partially addresses this tracker without closing it. It should be retargeted/rebased to `master` after #3528 lands. Remaining follow-up areas called out in the PR/docs include multiple-content availability, additional subtree metadata semantics, and tighter S2 child volumes.

**Comment by ibgreen** [comment]
Update: #3540 is now rebased onto current `master` (including the landed resource-intake work from #3528) and is in its final CI pass.

This means the lazy implicit-subtree expansion tranche is landing now: root tileset parsing no longer performs recursive subtree network I/O; traversal requests exactly one eligible visible/SSE-demanded subtree at a time; source-owned scheduling, deduplication, retry, and bounded metadata caching are in place; and the #3440 deepest-level / `availableLevels` boundary behavior is covered.

When #3540 merges, mark **Lazy implicit-subtree expansion** as complete and treat #3291 and #3440 as resolved by that PR. Keep this tracker open for multiple contents, metadata topology/semantics, dedicated skip-LOD traversal, cache-policy follow-up, S2 edge cases, and broader Cesium-conformance fixtures.

**Comment by ibgreen** [comment]
Opened draft PR #3667: https://github.com/visgl/loaders.gl/pull/3667. This tranche adds explicit multiple-content normalization and runtime loading, exposes all payloads through Tile3D.contents while preserving Tile3D.content, and adds the 3D Tiles format compatibility matrix. Implicit multiple-content availability and renderer-specific composition remain follow-ups.

**Comment by ibgreen** [comment]
Opened draft PR #3671: https://github.com/visgl/loaders.gl/pull/3671. This tranche preserves implicit subtree property-table, tile, content, and subtree metadata references on generated headers as implicitMetadata, with focused coverage. Metadata value/class decoding remains a follow-up.

**Comment by ibgreen** [comment]
Umbrella correctness PR #3846 is open. Its first tranche centralizes render-content visibility: multiple content volumes are treated as a union, clipping planes remain render-only, and contentless tiles fall back to their traversal volume. It also adds a conformance guide and focused browser tests. The tracker remains open; implicit subtree lifetime, dedicated skip-LOD, metadata semantics, and lifecycle fixtures remain planned follow-up tranches in the same umbrella effort.

**Comment by ibgreen** [comment]
## Post-correctness roadmap

The correctness foundation is now treated as complete in umbrella PR #3846: transform-aware geometric error, perspective/orthographic logical-pixel SSE, implicit availability arithmetic, content-volume union/clipping semantics, extension validation, metadata topology, and lifecycle diagnostics are covered.

The next tranches are:

| Name | Description | Impact | Complexity | Type |
|---|---|---|---|---|
| R1 Renderer contracts | Stable interfaces for multiple contents, feature IDs, metadata entities, styling inputs, and content visibility between tiles and deck.gl. | Very high | L | API / integration |
| R2 Styling and feature access | Renderer-neutral style inputs, property lookup, batch-table hierarchy queries, and conformance fixtures. | High | XL | Metadata / renderer |
| R3 Observability and benchmarks | Deterministic traversal snapshots, request/cache counters, stress fixtures, and performance budgets. | High | M | Performance / testing |
| R4 Fixed-frame coordinates | Consistent transforms and culling for local engineering datasets. | Medium–high | L | Coordinate systems |
| R5 Gaussian-splat tiles | KHR_gaussian_splatting and SPZ v2 decode paths coordinated with renderer support. | Strategic high | XL | Experimental format |
| R6 Vector/CAD tiles | EXT_mesh_polygon, 3DTILES_content_gltf_vector, and MVT-to-3D-Tiles conversion as standards stabilize. | Strategic medium–high | L | Experimental format / writer |
| R7 Voxel and terrain adapters | Experimental EXT_primitive_voxels and 3D Tiles-backed terrain source APIs. | Medium | XL | Experimental integration |

These replace the previous “remaining correctness” framing; the tracker stays open for renderer, integration, performance, and experimental work.

**Comment by ibgreen** [comment]
R1 is implemented in PR #4001: additive renderer-neutral contracts now expose ordered content entries, feature-ID declarations, raw metadata context, indexed/union content visibility, and lifecycle-safe payload state. The pure deck-layer adapter and API documentation are included. R2 styling/feature access can build on these contracts.

**Comment by ibgreen** [comment]
C3 is now open as PR #4013: https://github.com/visgl/loaders.gl/pull/4013

It hardens traversal and scheduler conformance with viewer-request-volume gating, external-tileset expiration handling, additive/replacement refinement coverage, and documentation of camera-motion safety. The change is renderer-neutral and preserves SSE targets.

**Comment by ibgreen** [comment]
C4 metadata tranche is now open as PR #4015: https://github.com/visgl/loaders.gl/pull/4015

This adds structural-metadata numeric normalization and property/class offset-scale transforms, plus a public `getStructuralMetadataRow` helper with `noData` and class-default handling. It is intentionally loader-side only; styling, picking, GPU upload, and renderer policy remain future work.

**Comment by ibgreen** [comment]
Opened #4018, “feat(3d-tiles): add traversal observability and benchmark baselines,” for roadmap R3. It adds deterministic Tileset3D traversal snapshots, request/cache/memory counters, implicit-subtree diagnostics, hermetic Vitest coverage, and benchmark guidance. R3 is now in review; R2 styling/feature access remains dependent on the renderer-neutral contracts and metadata work.

**Comment by ibgreen** [comment]
R2 styling/feature access tranche is in progress on `codex/3d-tiles-style-feature-access`: renderer-neutral style-input snapshots now merge tileset/group/tile/content metadata with hierarchy-aware batch-table rows, and named structural-metadata property access is exposed. Docs and conformance tests are included; styling-expression evaluation and GPU policy remain renderer-owned.

**Comment by ibgreen** [comment]
R2 implementation PR #4021 is open: https://github.com/visgl/loaders.gl/pull/4021. It adds renderer-neutral style-input snapshots, named structural-metadata property access, hierarchy-aware batch-table row composition, per-content group resolution, conformance tests, and the styling/feature-access guide. Styling expressions, picking policy, and GPU upload remain follow-up work.
